src/main.py

```diff
@@ -1836,155 +1829,142 @@ def _check_primary_switchover(self, db_state, zk_state):
-        limit = self.config.getfloat('global', 'postgres_timeout')
-        switchover_candidate = self._get_switchover_candidate()
+        logging.warning('Cluster was closed from user requests')
+
         if not helpers.await_for(
             lambda: self._candidate_is_sync_with_primary_with_get_state(switchover_candidate=switchover_candidate),
-            limit,
-            "replay lag become zero",
+            limit, "replay lag become zero",
         ):
             logging.error('check replica lsn diff failed - do not switchover')
             return False
 
-        # Store replics info
-        db_state = self.db.get_state()
-        if not self._store_replics_info(db_state, zk_state):
-            logging.error('replics_info was not stored - do not switchover')
+        logging.warning('Stopping postgresql (nowait)')
+        if self.db.stop_postgresql(wait=False, force_async=False) != 0:
+            logging.error('unable to stop postgresql')
             return False
 
-        # Announce intention to perform switchover to the rest of the cluster.
-        if not self.zk.write(self.zk.FAILOVER_INFO_PATH, 'switchover_initiated'):
-            logging.error(f'unable to write failover state to zk ({self.zk.FAILOVER_INFO_PATH})')
-            return False
+        # Give a sync replica good chance to catchup
+        # Note: we don't loose data here, as postgres stops in sync replication mode
+        time.sleep(5)
 
-        if not helpers.await_for(
-            lambda: self.zk.get(self.zk.SWITCHOVER_STATE_PATH) == "candidate_found", limit, "switchover candidate found"
-        ):
-            return False
-
-        # Attempt to shut down local PG instance.
-        # Failure is not critical.
-        if self.db.stop_postgresql(timeout=limit) == 0:
-            lsn = self._cmd_manager.get_control_parameter(db_state['pgdata'], "REDO location")
-            self.zk.noexcept_write(self.zk.SWITCHOVER_LSN_PATH, lsn)
-            if not self.zk.noexcept_write(self.zk.FAILOVER_INFO_PATH, 'switchover_master_shut'):
-                logging.error(f'unable to write failover state to zk ({self.zk.FAILOVER_INFO_PATH})')
-                return False
-        else:
-            logging.error('Unable to stop postgresql')
+        # this is the point of no-return for primary
+        # after that primary is stopped
+        if self._debug_failure('primary_switchover_before_release'):
             return False
 
         # Release leader-lock.
         # Wait 5 secs for the actual release.
+        logging.warning('Releasing the lock')
         self.zk.release_lock(lock_type=self.zk.PRIMARY_LOCK_PATH, wait=5)
 
+        logging.warning('Stopping postgresql (wait for complete)')
+        if self.db.stop_postgresql(timeout=limit, force_async=False) != 0:
+            if self.db.get_postgresql_status() == 0:
+                # pg is stopping, but still alive
+                logging.warning('unable to wait postgresql stopped')
```

Comment:
So, given we can't ensure that primary is shut down, how do we:
- convert this primary into replica?
- indicate that all WAL's have been successfully archived?
- prevent our other software (CMS) from decommissioning this host before WAL's are archived?